File: 145-binary-tree-postorder-traversal/145-binary-tree-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        Deque<TreeNode> s = new LinkedList<>();
        List<Integer> res = new ArrayList<>();

        TreeNode node = root;
        TreeNode prev = null;
        while(!s.isEmpty() || node != null) {
            while(node != null) {
                s.push(node);
                node = node.left;
            }
            node = s.pop();
            if(node.right == null || node.right == prev) {
                res.add(node.val);
                prev = node;
                node = null;
            }
            else {
                s.push(node);
                node = node.right;
            }
        }

        return res;
    }
}

File: 145-binary-tree-postorder-traversal/145-binary-tree-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        s1 = []
        s2 = []
        res = []

        if root: s1.append(root)
        while s1:
            head = s1.pop()
            s2.append(head.val)
            if head.left: s1.append(head.left)
            if head.right: s1.append(head.right)
        # while s2:
        #     res.append(s2.pop().val)

        s2.reverse()
        return s2